### PR TITLE
feat: expose session host route

### LINKED ISSUE #241: Prepare external reachable host names that auto setup route info

### PR DESCRIPTION
**Pull request by aslakknutsen**
Add a subdomain host to the gateway which targets a virtual service that auto-add routing headers. 
This allows a user to more easily call into a special route from e.g. a browser.  

fixes #241

### CODE REVIEW

File: e2e/smoke_test.go [hunk 1]
```diff
@@ -27,7 +27,8 @@ var _ = Describe("Smoke End To End Tests - against OpenShift Cluster with Istio
 		var (
 			namespace,
 			tmpDir string
-			scenario string
+			scenario    string
+			sessionname string
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply e2e/smoke_test.go]: https://github.com/Maistra/istio-workspace/pull/469/commits/43ec404c9d968eb2290ee1142d134c49c395d9b8

File: pkg/istio/virtualservice_test.go [hunk 2]
```diff
@@ -132,215 +265,350 @@ var _ = Describe("Operations for istio VirtualService kind", func() {
...
+											Destination: &istionetworkv1alpha3.Destination{
+												Host:   "details",
+												Subset: "v1-vs-test",
+											},
+										},
+									},
+									Match: []*istionetworkv1alpha3.HTTPMatchRequest{
+										&istionetworkv1alpha3.HTTPMatchRequest{
+											Headers: map[string]*istionetworkv1alpha3.StringMatch{
+												"x-test-suite": &istionetworkv1alpha3.StringMatch{MatchType: &istionetworkv1alpha3.StringMatch_Exact{Exact: "feature-x"}},
```
Comment (bartoszmajsak) [inline pkg/istio/virtualservice_test.go:349]: I recall you were showing me some "builders" for those objects? Or was that early stage of `Helpers`?

File: pkg/istio/virtualservicegatway_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var _ = Describe("Location of Gateway connected VirtualService Kind", func() {
+
+	// Missing tests:
+	//  - verify existing host logic from GW works (filter out our own sub domains)
```
Comment (bartoszmajsak) [inline pkg/istio/virtualservicegatway_test.go:23]: Is that `// TODO` for this PR?

File: pkg/k8s/deployment_test.go [hunk 4]
```diff
@@ -37,12 +42,14 @@ var _ = Describe("Operations for k8s Deployment kind", func() {
 		schema := runtime.NewScheme()
 		err := appsv1.AddToScheme(schema)
 		Expect(err).ToNot(HaveOccurred())
+		c = fake.NewFakeClientWithScheme(schema, objects...)
+		get = operator.New(c)
```
Comment (bartoszmajsak) [inline pkg/k8s/deployment_test.go:46]: I have a bit of a problem understanding this. You are creating new operator called `get`?

File: pkg/model/model.go [hunk 5]
```diff
@@ -42,15 +42,62 @@ type HostName struct {
 	Namespace string
 }
 
-// GetTargetsByKind returns the targets of the given kind.
-func (r *Ref) GetTargetsByKind(kinds ...string) []LocatedResourceStatus {
+// Predicate base function to filter Resources.
+type Predicate func(ResourceStatus) bool
```
Comment (bartoszmajsak) [inline pkg/model/model.go:46]: This is nice!

File: test/operator/getters.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		VirtualServices:           VirtualServices(c),
+		Deployment:                Deployment(c),
+		DeploymentWithError:       DeploymentWithError(c),
+		DeploymentConfig:          DeploymentConfig(c),
+		DeploymentConfigWithError: DeploymentConfigWithError(c),
+	}
+}
+
+// Helpers simple struct to hold funcs.
+type Helpers struct {
```
Comment (bartoszmajsak) [inline test/operator/getters.go:36]: I think we should work a bit on the naming here. Is it operator, helper, getter, or bunch of builders?

File: pkg/controller/session/session_controller_int_test.go [hunk 7]
```diff
@@ -90,32 +79,165 @@ var _ = Describe("Complete session manipulation", func() {
...
+								"image": "x:x:x",
+							},
+						},
+					},
+				},
+			})
 		})
 
-		It("with update", func() {
+		It("with ref update", func() {
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_int_test.go:101]: The IT here is more a "Under what circumstance". The Assertion is more a "That's something that should have changed" more than we're ensuring that changed I guess. The double Reconcile call is; First call "Create", (update the ref config), Second call "Update".  It essentially simulates these two calls: 

* ike create -s x -d a --image x
* ike create -s x -d b --image y

I'll add some more description to it

File: pkg/controller/session/session_controller_int_test.go [hunk 8]
```diff
@@ -90,32 +79,165 @@ var _ = Describe("Complete session manipulation", func() {
...
+			c.Update(context.Background(), &session)
+
+			res1, err = controller.Reconcile(req1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(res2.Requeue).To(BeFalse())
+
+			// Verify - no vs removed (only gateway connected duplicated)
+			vss = get.VirtualServices("test")
+			for _, vs := range vss.Items {
+				fmt.Println(vs.Name)
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_int_test.go:224]: leftover :)

File: pkg/istio/virtualservice_test.go (hunk 2 above)
Comment (aslakknutsen) [reply pkg/istio/virtualservice_test.go:349]: We have the scenario generators but that's different. I guess we could expand on them to use them here too.

The `Helpers` are just "Type-safe" wrappers for getting RuntimeObjects from the Client.

File: pkg/k8s/deployment_test.go (hunk 4 above)
Comment (aslakknutsen) [reply pkg/k8s/deployment_test.go:46]: The naming of the "operator" test helper package is a bit nonhelpful in the construct call, I agree. "getters", "helpers", not sure what to call it

File: test/operator/getters.go (hunk 6 above)
Comment (aslakknutsen) [reply test/operator/getters.go:36]: They are helpers for the test suite that makes it easier/cleaner to get objects from the client. They don't really build anything.

File: e2e/smoke_test.go [hunk 9]
```diff
@@ -40,6 +41,7 @@ var _ = Describe("Smoke End To End Tests - against OpenShift Cluster with Istio
 			EnablePullingImages(namespace)
 			InstallLocalOperator(namespace)
 			DeployTestScenario(scenario, namespace)
+			sessionName = GenerateSessionName(7)
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply e2e/smoke_test.go]: https://github.com/maistra/istio-workspace/pull/469/commits/ad76ceb174c1294a2b270cdd5e91877d4df45505

File: pkg/controller/session/session_controller_int_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_int_test.go]: Better? https://github.com/maistra/istio-workspace/pull/469/commits/281fefc7610d7265159640ea9703e04f18ef2722

File: pkg/controller/session/session_controller_int_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_int_test.go]: https://github.com/maistra/istio-workspace/pull/469/commits/e5bc6c426a4197a92fedba94514370159057aa91

File: pkg/istio/virtualservicegatway_test.go (hunk 3 above)
Comment (aslakknutsen) [reply pkg/istio/virtualservicegatway_test.go]: https://github.com/maistra/istio-workspace/pull/469/commits/f006ece2d6ac587e091b04cd96c7c726fa1cecd3

File: pkg/k8s/deployment_test.go (hunk 4 above)
Comment (aslakknutsen) [reply pkg/k8s/deployment_test.go]: https://github.com/maistra/istio-workspace/pull/469/commits/597ac6def9ade65dd8ef08a58a84741ada556acf

File: test/operator/getters.go (hunk 6 above)
Comment (aslakknutsen) [reply test/operator/getters.go]: https://github.com/maistra/istio-workspace/pull/469/commits/597ac6def9ade65dd8ef08a58a84741ada556acf

**Pull request merged**